Add unreleased notes and documentation on the constraints section (#9845)
crate/crate: docs/sql/general/constraints.rst

For further details about the meaning of the options see
:ref:`indices_and_fulltext`.

.. _check_constraint:

``CHECK``
---------

The CHECK constraint specifies that the value/s of certain column/s must satisfy
a boolean expression on insert and update.

Syntax:

::

  [CONSTRAINT <name>] CHECK (boolean_expression)

If `CONSTAINT <name>` is omitted, CrateDB generates a unique name automatically.

> This name is visible in :ref:`information_schema` and can be used to

Reviewer: please correct the `ref` to point to the relevant table.. sorry I was too lazy in the suggestion to look it up.

:ref:`ALTER TABLE DROP` a constraint.

::

  cr> select table_schema, table_name, constraint_type, constraint_name
  ...     from information_schema.table_constraints
  ...     where table_name = 'metrics1';

The CONSTRAINT definition can either be inline with a column, like this::

    cr> CREATE TABLE metrics1 (
    ...     weight REAL CONSTRAINT weight_is_positive CHECK (weight >= 0),
    ... );

Or, also inline, but without explicit name::

    cr> CREATE TABLE metrics2 (
    ...     weight REAL CHECK (weight >= 0),

>     ... );

Reviewer: The examples should be runnable and have the outputs asserted, no?

> Or, on a table level with explicit name:

Reviewer: ```suggestion
Or, on a table level with explicit name::
```

    cr> CREATE TABLE metrics3 (
    ...     weight REAL,
    ...     CONSTRAINT weight_is_positive CHECK (weight >= 0),
    ... );

> Or without name:

Reviewer: ```suggestion
Or without name::
```

    cr> CREATE TABLE metrics4 (
    ...     weight REAL,
    ...     CHECK (weight >= 0),
    ... );

Multiple columns can be referenced:

    cr> CREATE TABLE metrics5 (
    ...     weight REAL,
    ...     qty INTEGER,
    ...     CHECK (weight * qty != 1918),
    ... );

>     cr> CREATE TABLE metrics6 (
>     ...     weight REAL,
>     ...     qty INTEGER CHECK (qty * weight != 1918)
>     ... );
> 
>     cr> CREATE TABLE metrics7 (
>     ...     weight REAL,
>     ...     qty INTEGER,
>     ...     CHECK (qty * weight != 1918)
>     ... );

Reviewer: ```suggestion
```

I think one example is enough to illustrate that point. I don't think we need to show all permutations.

For further details about the meaning of the options see
:ref:`indices_and_fulltext`.

.. _check_constraint:

``CHECK``
---------

The CHECK constraint specifies that the value/s of certain column/s must satisfy
a boolean expression on insert and update.

Syntax:

::

>   check(boolean_expression)
>   constraint <name> check(boolean_expression)

Reviewer: ```suggestion
  [CONSTRAINT <name>] CHECK (boolean_expression)
```

> With two options:
> 
> 1) As many check constraints as required, as additional elements of
>    a create table statement.
>    When name is omitted, one is generated with pattern `fqtn_check_uuid`,
>    where:
> 
>    - `fqtb`: full qualified table name.
>    - `check`: "check".
>    - `uuid`: 6 random hexadecimal digits (2 octets),
>      e.g. doc_my_table_check_bd3f57640937.
> 
> 2) A single check constraint is allowed as part of a column definition.
>    When name is omitted, the pattern is `fqtn_column_check_uuid`,
>    where:
> 
>    - `column`: column name.
> 
> Example:

Reviewer: I think this is a bit hard to understand. How about:

```suggestion
If `CONSTAINT <name>` is omitted, CrateDB generates a unique name automatically. This name is visible in :ref:`.. ref to information_schema table` and can be used to :ref:`ALTER TABLE DROP` a constraint.
```

Leave out the implementation details of the name generation. (Not that interesting, and allows us to change it later on - as we can treat this as an implementation detail).

```
The CONSTRAINT definition can either be inline with a column, like this::

    cr> CREATE TABLE metrics1 (
    ...     weight REAL CONSTRAINT weight_is_positive CHECK (weight >= 0),
    ... );

Or, also inline, but without explicit name::

    cr> CREATE TABLE metrics2 (
    ...     weight REAL CHECK (weight >= 0),
    ... );

Or on a table level, in which case the `CONSTRAINT` keyword and constraint name is required:

    cr> CREATE TABLE metrics3 (
    ...     weight REAL,
    ...     CONSTRAINT weight_is_positive CHECK (weight >= 0),
    ... );

 ```

Maybe another example which points out that multiple columns can be referenced.

::

  cr> create table my_table (
  ...    first_column integer primary key,
  ...    second_column integer check(second_column >= 0),
  ...    third_column string constraint not_you check(third_column <> 'me'),
  ...    check(first_column >= 100),
  ...    constraint is_even check(second_column % 2 = 0)
  ...);
  CREATE OK, 1 row affected  (... sec)

::

  cr> select table_schema, table_name, constraint_type, constraint_name
  ...     from information_schema.table_constraints
  ...     where table_name = 'my_table';

.. WARNING::

>    CHECK constraint conditions must be immutable, always giving the same

Reviewer: ```suggestion
   CHECK constraint conditions must be deterministic, always yielding the same
```

   result for the same input.
   A way to break this is to reference a user-defined function in a CHECK
   expression, and then change the behavior of that function. Some existing
   rows in the table could now violate the CHECK constraint. That would

>    cause a subsequent database dump and reload to fail.

Reviewer: Not sure if we should include hints how to break the system.
